R/rows.R

```diff
@@ -0,0 +1,610 @@
+  out <- attr(x, "returned_rows", TRUE)
+  if (is.null(out)) {
+    abort("No returned rows available.")
+  }
+  out
+}
+
+#' has_returned_rows()
+#'
+#' `has_returned_rows()` checks if `x` has stored RETURNING rows produced by
+#' [rows_insert()], [rows_update()], [rows_upsert()], or [rows_delete()].
+#'
+#' @param x A lazy tbl.
+#'
+#' @return For `has_returned_rows()`, a scalar logical.
+#'
+#' @rdname get_returned_rows
+#' @export
+has_returned_rows <- function(x) {
+  !identical(attr(x, "returned_rows"), NULL)
+}
+
+#' @export
+sql_returning_cols.duckdb_connection <- function(con, cols, ...) {
+  abort("DuckDB does not support the `returning` argument.")
+}
+
+sql_coalesce <- function(x, y) {
+  sql(paste0("COALESCE(", x, ", ", y, ")"))
+}
+
+# check helpers -----------------------------------------------------------
+
+rows_check_by <- function(by, y, ..., error_call = caller_env()) {
+  check_dots_empty()
+
+  if (is.null(by)) {
+    if (ncol(y) == 0L) {
+      abort("`y` must have at least one column.", call = error_call)
+    }
+
+    by <- colnames(y)[[1]]
+
+    inform(
+      message = glue("Matching, by = \"{by}\""),
+      class = c("dplyr_message_matching_by", "dplyr_message")
+    )
+  }
+
+  if (!is.character(by)) {
+    abort("`by` must be a character vector.", call = error_call)
+  }
+  if (is_empty(by)) {
+    abort("`by` must specify at least 1 column.", call = error_call)
+  }
+  if (!all(names2(by) == "")) {
+    abort("`by` must be unnamed.", call = error_call)
+  }
+
+  by
+}
+
+rows_check_containment <- function(x, y, ..., error_call = caller_env()) {
+  check_dots_empty()
+
+  bad <- setdiff(colnames(y), colnames(x))
+
+  if (!is_empty(bad)) {
+    bad <- err_vars(bad)
+
+    message <- c(
+      "All columns in `y` must exist in `x`.",
+      i = glue("The following columns only exist in `y`: {bad}.")
+    )
+
+    abort(message, call = error_call)
+  }
+
+  invisible()
+}
+
+rows_check_key <- function(x,
+                           by,
+                           arg,
+                           ...,
+                           unique = FALSE,
+                           error_call = caller_env()) {
+  check_dots_empty()
+
+  missing <- setdiff(by, colnames(x))
+
+  if (!is_empty(missing)) {
+    missing <- err_vars(missing)
+
+    message <- c(
+      "All columns specified through `by` must exist in `x` and `y`.",
+      i = glue("The following columns are missing from `{arg}`: {missing}.")
+    )
+
+    abort(message, call = error_call)
+  }
+
+  # TODO adapt this
+  # out <- x[by]
+  #
+  # if (unique && vec_duplicate_any(out)) {
+  #   duplicated <- vec_duplicate_detect(out)
+  #   duplicated <- which(duplicated)
+  #   duplicated <- err_locs(duplicated)
+  #
+  #   message <- c(
+  #     glue("`{arg}` key values must be unique."),
+  #     i = glue("The following rows contain duplicate key values: {duplicated}.")
+  #   )
+  #
+  #   abort(message, call = error_call)
+  # }
+  #
+  # out
```

Comment:
I don't think we should be doing anything here. We have tools in dm to check referential integrity and key constraints "after the fact", and we can simulate with `in_place = FALSE` . Actively checking here can be very expensive.

Reply:
What do you mean by "after the fact"?

There should definitely be tools to easiliy check the uniqueness of `y` and they should be well-documented.

If `y` is a local table the check is easy, so it would be an option to check this before copying.